pkgs/development/libraries/oracle-instantclient/default.nix

```diff
@@ -24,37 +26,44 @@ let
       (requireSource "basic" arch version "" "5015e3c9fba84e009f7519893f798a1622c37d1ae2c55104ff502c52a0fe5194")
       (requireSource "sdk" arch version "" "7f404c3573c062ce487a51ac4cfe650c878d7edf8e73b364ec852645ed1098cb")
       (requireSource "sqlplus" arch version "" "d49b2bd97376591ca07e7a836278933c3f251875c215044feac73ba9f451dfc2") ]
-      ++ stdenv.lib.optional odbcSupport (requireSource "odbc" arch version "2" "365a4ae32c7062d9fbc3fb41add748e7881f774484a175a4b41a2c294ce9095d");
+      ++ optional odbcSupport (requireSource "odbc" arch version "2" "365a4ae32c7062d9fbc3fb41add748e7881f774484a175a4b41a2c294ce9095d");
     "x86_64-darwin" = [
       (requireSource "basic" arch version "2" "3ed3102e5a24f0da638694191edb34933309fb472eb1df21ad5c86eedac3ebb9")
       (requireSource "sdk" arch version "2" "e0befca9c4e71ebc9f444957ffa70f01aeeec5976ea27c40406471b04c34848b")
       (requireSource "sqlplus" arch version "2" "d147cbb5b2a954fdcb4b642df4f0bd1153fd56e0f56e7fa301601b4f7e2abe0e") ]
-      ++ stdenv.lib.optional odbcSupport (requireSource "odbc" arch version "2" "1805c1ab6c8c5e8df7bdcc35d7f2b94c329ecf4dff9bde55d5f9b159ecd8b64e");
+      ++ optional odbcSupport (requireSource "odbc" arch version "2" "1805c1ab6c8c5e8df7bdcc35d7f2b94c329ecf4dff9bde55d5f9b159ecd8b64e");
   }."${stdenv.system}" or throwSystem;
 
+  extLib = stdenv.hostPlatform.extensions.sharedLibrary;
 in stdenv.mkDerivation rec {
   inherit version srcs;
   name = "oracle-instantclient-${version}";
 
   buildInputs = [ stdenv.cc.cc.lib ]
-    ++ stdenv.lib.optionals (stdenv.isLinux) [ libaio ]
-    ++ stdenv.lib.optional odbcSupport unixODBC;
-  nativeBuildInputs = [ autoPatchelfHook makeWrapper unzip ];
+    ++ optionals (stdenv.isLinux) [ libaio ]
+    ++ optional odbcSupport unixODBC;
+  nativeBuildInputs = [ autoPatchelfHook makeWrapper unzip ] ++ optional stdenv.isDarwin fixDarwinDylibNames;
 
   unpackCmd = "unzip $curSrc";
 
   installPhase = ''
     mkdir -p "$out/"{bin,include,lib,"share/java","share/${name}/demo/"}
-
     install -Dm755 {sqlplus,adrci,genezi} $out/bin
+    ${optionalString stdenv.isDarwin ''
+      for exe in "$out/bin/"* ; do
+        install_name_tool -add_rpath "$out/lib" "$exe"
```

Comment:
The oracle instantclient binary zip sources are packaged as a flat structure and contain an RPATH relative to the executable (e.g. sqlplus, etc), so the embedded load commands don't find the associated libs when installed in the nix store under ./bin, ./lib tree structure. For example listing the `sqlplus` binary in the `instantclient-sqlplus-linux.x64-12.2.0.1.0.zip` source gives this:
```
$ otool -l sqlplus
Load command 12
          cmd LC_LOAD_DYLIB
      cmdsize 48
         name @rpath/libsqlplus.dylib (offset 24)
   time stamp 2 Wed Dec 31 17:00:02 1969
      current version 0.0.0
compatibility version 0.0.0
Load command 13
          cmd LC_LOAD_DYLIB
      cmdsize 56
         name @rpath/libclntsh.dylib.12.1 (offset 24)
   time stamp 2 Wed Dec 31 17:00:02 1969
      current version 0.0.0
compatibility version 0.0.0
Load command 14
          cmd LC_LOAD_DYLIB
      cmdsize 56
         name @rpath/libclntshcore.dylib.12.1 (offset 24)
   time stamp 2 Wed Dec 31 17:00:02 1969
      current version 0.0.0
compatibility version 0.0.0
Load command 15
          cmd LC_LOAD_DYLIB
      cmdsize 48
         name @rpath/libnnz12.dylib (offset 24)
   time stamp 2 Wed Dec 31 17:00:02 1969
      current version 0.0.0
compatibility version 0.0.0
Load command 18
          cmd LC_RPATH
      cmdsize 32
         path @executable_path/ (offset 12)
```

I would have thought `fixDarwinDylibNames` would work on executables similar to how `autoPatchelfHook` works, but digging into the source reveals that function only works on `*.dylib` files. Adding the call to `install_name_tool` was the only way I could get the `sqlplus` and other executables to run without having to mess around with setting the LD_LIBRARY_PATH or DYLD_LIBRARY_PATH in the environment. I am definitely a nix newb, so I'm not sure if there is a different function than `fixDarwinDylibNames` that works like `autoPatchelfHook`. I couldn't find anything in the standard nixpkgs setup-hooks.

Reply:
I think it's fine adding `$out/lib` to the rpath of all binaries - we do similar things in other places in Nixpkgs as well.

Reply:
@mikemckibben I see, changing the LC_LOAD_DYLIB to use absolute paths instead would be nicer IMHO but this works. `fixDarwinDylibNames` does indeed only handle dylibs and since this isn't built from source we can't fix the rpath referrences before the binaries are linked against the libraries, that would have made this unnecessary.